back/spec/services/side_fx_group_spec.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe SideFxGroupService do
+  let(:service) { described_class.new }
+  let(:current_user) { create(:user) }
+  let(:group) { create(:group) }
+
+  describe 'after_update' do
+    it "logs a 'changed' action job when the record has been updated" do
+      travel_to(Time.now) do # rubocop:disable Rails/FreezeTime
+        original_title = group.title_multiloc
+        group.update(title_multiloc: { 'en' => 'Updated Title' })
+
+        expect { service.after_update(group, current_user) }
+          .to have_enqueued_job(LogActivityJob).with(
+            [group.class.name, group.id].join('/'), # encode_frozen_resource(group),
+            'changed',
+            current_user,
+            Time.now.to_i,
+            payload: { changes: { 'title_multiloc' => [original_title, { 'en' => 'Updated Title' }] } }
+          )
+      end
+    end
+
+    it 'corrrectly logs update payload even when original record changed again before logging job runs' do
+      travel_to(Time.now) do # rubocop:disable Rails/FreezeTime
+        original_title = group.title_multiloc
+        group.update(title_multiloc: { 'en' => 'Updated Title' })
+
+        service.after_update(group, current_user)
+
+        # group.update(title_multiloc: { 'en' => 'Updated Title again' })
+
+        # This is probably happening after the job is executed, so does not really test what I want to test.
+        group.destroy!
+
+        expect(LogActivityJob).to(have_been_enqueued.at_least(:once).with.with(
+          [group.class.name, group.id].join('/'), # encode_frozen_resource(group),
+          'changed',
+          current_user,
+          Time.now.to_i,
+          payload: { changes: { 'title_multiloc' => [original_title, { 'en' => 'Updated Title' }] } }
+        ))
+      end
+    end
```

Comment:
I am trying to test what happens if the group is updated (again) or deleted _after_ the initial `LogActivityJob` to log the update is enqueued _and before_ the job is processed.

I don’t think this test really tests this, and I couldn’t find an example test that does this (not anything online about how to do this). Any ideas?

Reply:
That would indeed be hard to test, as you need some kind of asynchronous job processing, which we don't have in tests. One strategy would be to cover this in unit tests on both sides, the side_fx service and the actual job. You wouldn't be testing that information is properly passed between them, so then you'd take proper care to make sure those are aligned between the 2 (people sometimes refer to this as "the contract", and its something you accept to be fixed from a testing perspective).

In that case, though, I wouldn't bother. The moment you call `previous_changes` is the moment that function executes and ActiveRecord returns the `changes` data structure. We then schedule it as a job, so it's written out in that form to the jobs table. In other words, you can be sure than any data manipulation after will _not_ be covered in the changes you enqueued before.